### PR TITLE
Refactor CI workflow

### PR DESCRIPTION
**Pull request by jgarber623**
## Description

This PR renames and refactors the project's GitHub Actions CI workflow. The updated workflow will now run on:

- pushes to `main`,
- pull requests against `main`, and
- by manual invocation via GitHub's UI

Notable changes to the workflow include:

- use [CargoSense/setup-elixir-project](https://github.com/CargoSense/setup-elixir-project) reusable workflow (in Ubuntu job, at least)
- ~~adds Ubuntu and macOS to the test matrix~~
- refactors existing Ubuntu test job
- adds a new macOS test job
- specifies oldest and newest Elixir and Erlang combo in the matrix
- use latest [actions/checkout](https://github.com/actions/checkout)
- sets job name to use matrix values for clarity

The macOS job takes a necessarily different tact for installing Erlang and Elixir owing to the underlying `erlef/setup-beam` workflow not (yet) supporting macOS (see https://github.com/erlef/setup-beam/issues/54).

### CODE REVIEW

**Comment by jgarber623** [comment]
Error on the macOS job:

> Error: @erlef/setup-beam only supports Ubuntu and Windows at this time

😭

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -1,53 +0,0 @@
...
-      - run: mix deps.compile
-
-      - run: mix compile --warnings-as-errors
-        if: ${{ matrix.lint }}
-
-      - run: mix test
-        if: ${{ ! matrix.lint }}
-
-      - run: mix test --warnings-as-errors
-        if: ${{ matrix.lint }}
```
Comment (jgarber623) [inline .github/workflows/main.yml:53]: @mcrumm Showing my own lack of knowledge here, but I'm not entirely sure I've captured all the test cases in the updated YAML file.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        run: |
+          echo "elixir ${{ matrix.pair.elixir }}" >> .tool-versions
+          echo "erlang ${{ matrix.pair.otp }}" >> .tool-versions
+          cat .tool-versions
+      - uses: asdf-vm/actions/install@v3
+      - name: Install Hex package manager
+        run: mix local.hex --force
+      - name: Install dependencies
+        run: mix deps.get
+      - run: mix format --check-formatted
```
Comment (mcrumm) [inline .github/workflows/ci.yml:37]: No need to run the formatter on macOS, we really only need to run it against the latest Elixir/OTP.

```suggestion
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            otp: "27.0.1"
+    steps:
+      - uses: actions/checkout@v4
+      - uses: CargoSense/setup-elixir-project@v1
+        with:
+          otp-version: ${{ matrix.pair.otp }}
+          elixir-version: ${{ matrix.pair.elixir }}
+          build-flags: "--warnings-as-errors"
+      - run: mix format --check-formatted
+      - run: mix test
```
Comment (mcrumm) [inline .github/workflows/ci.yml:58]: The `lint` portion of the matrix is there to ensure certain command/option combinations (e.g. `mix format`, `mix test --warnings-as-errors`) are run only on the latest Elixir/OTP pair.

```suggestion
      matrix:
        include:
          - pair:
              elixir: "1.13"
              otp: "24.3.4.10"
          - pair:
              elixir: "1.17"
              otp: "27.0.1"
            lint: lint
    steps:
      - uses: actions/checkout@v4
      - uses: erlef/setup-beam@main
        with:
          otp-version: ${{ matrix.pair.otp }}
          elixir-version: ${{ matrix.pair.elixir }}
          version-type: strict
      - uses: actions/cache@v4
        with:
          path: deps
          key: mix-deps-${{ hashFiles('**/mix.lock') }}
      - run: mix deps.get --check-locked
      - run: mix format --check-formatted
        if: ${{ matrix.lint }}
      - run: mix deps.unlock --check-unused
        if: ${{ matrix.lint }}
      - run: mix deps.compile
      - run: mix compile --warnings-as-errors
        if: ${{ matrix.lint }}
      - run: mix test
        if: ${{ ! matrix.lint }}
      - run: mix test --warnings-as-errors
        if: ${{ matrix.lint }}
```

Note I'm also recommending we skip our custom GitHub Action in favor of `erlef/setup-beam`. The phoenix_playground repo has a simplified config that should work fine for this repo:
- https://github.com/phoenix-playground/phoenix_playground/blob/main/.github/workflows/ci.yml#L1

**Review by mcrumm — APPROVED** [review]
macOS runners ftw! 🚀

**Pull request merged**